flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/EdgeManagerBuildUtil.java

```diff
@@ -0,0 +1,174 @@
+            IntermediateResult ires,
+            int inputNumber,
+            DistributionPattern distributionPattern) {
+
+        switch (distributionPattern) {
+            case POINTWISE:
+                connectPointwise(taskVertices, ires, inputNumber);
+                break;
+            case ALL_TO_ALL:
+                connectAllToAll(taskVertices, ires, inputNumber);
+                break;
+            default:
+                throw new RuntimeException("Unrecognized distribution pattern.");
+        }
+    }
+
+    private static void connectAllToAll(
+            ExecutionVertex[] taskVertices, IntermediateResult ires, int inputNumber) {
+
+        ConsumedPartitionGroup consumedPartitions =
+                new ConsumedPartitionGroup(
+                        Arrays.stream(ires.getPartitions())
+                                .map(IntermediateResultPartition::getPartitionId)
+                                .collect(Collectors.toList()));
+        for (ExecutionVertex ev : taskVertices) {
+            ev.setConsumedPartitions(consumedPartitions, inputNumber);
+        }
+
+        ConsumerVertexGroup vertices =
+                new ConsumerVertexGroup(
+                        Arrays.stream(taskVertices)
+                                .map(ExecutionVertex::getID)
+                                .collect(Collectors.toList()));
+        for (IntermediateResultPartition partition : ires.getPartitions()) {
+            partition.setConsumers(vertices);
+        }
+    }
+
+    private static void connectPointwise(
+            ExecutionVertex[] taskVertices, IntermediateResult ires, int inputNumber) {
+
+        final int sourceCount = ires.getPartitions().length;
+        final int targetCount = taskVertices.length;
+
+        if (sourceCount == targetCount) {
+            for (int i = 0; i < sourceCount; i++) {
+                ExecutionVertex executionVertex = taskVertices[i];
+                IntermediateResultPartition partition = ires.getPartitions()[i];
+
+                ConsumerVertexGroup consumerVertexGroup =
+                        new ConsumerVertexGroup(executionVertex.getID());
+                partition.setConsumers(consumerVertexGroup);
+
+                ConsumedPartitionGroup consumedPartitionGroup =
+                        new ConsumedPartitionGroup(partition.getPartitionId());
+                executionVertex.setConsumedPartitions(consumedPartitionGroup, inputNumber);
+            }
+        } else if (sourceCount > targetCount) {
+            for (int index = 0; index < targetCount; index++) {
+
+                ExecutionVertex executionVertex = taskVertices[index];
+                ConsumerVertexGroup consumerVertexGroup =
+                        new ConsumerVertexGroup(executionVertex.getID());
+
+                List<IntermediateResultPartitionID> consumedPartitions =
+                        new ArrayList<>(sourceCount / targetCount + 1);
+
+                if (sourceCount % targetCount == 0) {
```

Comment:
This if-else looks unnecessarily complex.
I think it can be simplified as below. WDYT?

```
                int start = index * sourceCount / targetCount;
                int end = (index + 1) * sourceCount / targetCount;
                List<IntermediateResultPartitionID> consumedPartitions =
                        new ArrayList<>(end - start);
                for (int i = start; i < end; i++) {
                    IntermediateResultPartition partition = ires.getPartitions()[i];
                    partition.setConsumers(consumerVertexGroup);
                    consumedPartitions.add(partition.getPartitionId());
                }
```

Reply:
Totally agreed. Thanks for providing a better solution. I've validate it with several common cases. I'm wondering that could we simplify the situation `sourceCount > targetCount` in this way...?

Reply:
Yes I think the `sourceCount > targetCount` case can be simplified similarly.
However, I noticed that in `PointwisePatternTest` it does not verify whether the way to connection is changed in irregular case. I prefer to not change how edges are connected even if the connections are still evenly distributed, i.e. `inEdges.length >= factor && inEdges.length <= factor + delta` in `PointwisePatternTest#testHighToLow`.

So I think it's better to add some case like `6 to 4`, `4 to 6`, `5 to 3`, `3 to 5` in `PointwisePatternTest` and verify the connections are exactly as what it was.

Reply:
Thanks for the suggestion. I've improved the `PointwisePatternTest` and the logics in `EdgeManagerBuilUtil`.

Reply:
@zhuzhurk Thanks for the suggestion. `PointwisePatternTest` has been improved.